nemo/collections/asr/parts/utils/online_clustering.py

```diff
@@ -0,0 +1,1106 @@
+
+    def _init_memory_buffer_variables(self):
+        """
+        Initialize memory buffer related variables.
+
+        Attributes:
+            max_embed_count (int):
+                The maximum number of segments the streaming system has ever seen
+            memory_margin (int):
+                The margin that is added to keep the segmentation data in the streaming system
+            _minimum_segments_per_buffer (int):
+                Maximum number of embedding vectors kept in history buffer per speaker.
+                Example:
+                    history_buffer_size (history_n) = 100
+                    max_num_speakers = 4
+                    _minimum_segments_per_buffer = 25
+            history_buffer_seg_end (int):
+                Index that indicates the boundary between history embedding sets and current processing buffer
+                when history embedding vectors and current input embedding vectors are concatenated into a
+                single matrix.
+        """
+        self.max_embed_count = 0
+        self.memory_margin = 0
+        self._minimum_segments_per_buffer = int(self.history_n / self.max_num_speakers)
+        self.history_buffer_seg_end = 0
+
+    def _init_memory_embeddings(self):
+        """
+        Initialize history buffer related variables.
+
+        Attributes:
+            is_online (bool):
+                If self.is_online is False:
+                    FIFO queue does not push out any speaker embedding vector
+                If self.is_online is True:
+                    FIFO queue starts push out speaker embedding vectors and saving them into
+                    history buffer.
+            history_embedding_buffer_emb (Tensor)
+                Tensor containing speaker embedding vectors for saving the history of the previous
+                speaker profile in the given audio session
+            history_embedding_buffer_label (Tensor)
+                Speaker label (cluster label) for embedding vectors saved in the history buffer
+            Y_fullhist (Tensor)
+                Tensor containing the speaker label hypothesis from start to current frame
+        """
+        self.is_online = False
+        self.history_embedding_buffer_emb = torch.tensor([])
+        self.history_embedding_buffer_label = torch.tensor([])
+        self.Y_fullhist = torch.tensor([])
+
+    def onlineNMEanalysis(self, nmesc: NMESC, frame_index: int) -> Tuple[torch.LongTensor, torch.Tensor]:
+        """
+        To save the running time, the p-value is only estimated in the beginning of the session.
+        After switching to online mode, the system uses the most common estimated p-value.
+        Estimating p-value requires a plenty of computational resource. The less frequent estimation of
+        p-value can speed up the clustering algorithm by a huge margin.
+
+        Args:
+            nmesc: (NMESC)
+                nmesc instance.
+            is_online: (bool)
```

Comment:
I don;t see this in the argument and frame_index definition is missing

Reply:
fixed this docstring